// firebase-tools/lib/firebase_studio/migrate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractMetadata = extractMetadata;
exports.uploadSecrets = uploadSecrets;
exports.migrate = migrate;
const fs = require("fs/promises");
const path = require("path");
const child_process_1 = require("child_process");
const semver = require("semver");
const logger_1 = require("../logger");
const prompt = require("../prompt");
const apphosting = require("../gcp/apphosting");
const utils = require("../utils");
const templates_1 = require("../templates");
const track = require("../track");
const secrets_1 = require("../apphosting/secrets");
const env = require("../functions/env");
const error_1 = require("../error");
const os = require("os");
const agentSkills_1 = require("../agentSkills");
async function setupAntigravityMcpServer(rootPath, appType, nonInteractive) {
    const mcpConfigDir = path.join(os.homedir(), ".gemini", "antigravity");
    const mcpConfigPath = path.join(mcpConfigDir, "mcp_config.json");
    let mcpConfig = { mcpServers: {} };
    try {
        await fs.mkdir(mcpConfigDir, { recursive: true });
        const content = await fs
            .readFile(mcpConfigPath, "utf-8")
            .catch((err) => {
            if (err.code === "ENOENT") {
                return null;
            }
            throw err;
        });
        if (content) {
            mcpConfig = JSON.parse(content);
            if (!mcpConfig.mcpServers) {
                mcpConfig.mcpServers = {};
            }
        }
        let updated = false;
        if (!mcpConfig.mcpServers["firebase"]) {
            if (utils.commandExistsSync("npx")) {
                const confirmFirebase = await prompt.confirm({
                    message: "Would you like to enable the Firebase MCP server for Antigravity?",
                    default: true,
                    nonInteractive,
                });
                if (confirmFirebase) {
                    mcpConfig.mcpServers["firebase"] = {
                        command: "npx",
                        args: ["-y", "firebase-tools@latest", "mcp", "--dir", path.resolve(rootPath)],
                    };
                    updated = true;
                    logger_1.logger.info(`✅ Configured Firebase MCP server in ${mcpConfigPath}`);
                }
            }
            else {
                logger_1.logger.info("ℹ️ npx not found on PATH, skipping Firebase MCP server configuration.");
            }
        }
        else {
            logger_1.logger.info("ℹ️ Firebase MCP server already configured in Antigravity, skipping.");
        }
        if (appType === "FLUTTER") {
            if (utils.commandExistsSync("dart")) {
                if (!mcpConfig.mcpServers["dart"]) {
                    const confirmDart = await prompt.confirm({
                        message: "Would you like to enable the Dart MCP server for Antigravity?",
                        default: true,
                        nonInteractive,
                    });
                    if (confirmDart) {
                        mcpConfig.mcpServers["dart"] = {
                            command: "dart",
                            args: ["mcp-server"],
                        };
                        updated = true;
                        logger_1.logger.info(`✅ Configured Dart MCP server in ${mcpConfigPath}`);
                    }
                }
                else {
                    logger_1.logger.info("ℹ️ Dart MCP server already configured in Antigravity, skipping.");
                }
            }
            else {
                utils.logWarning("Couldn't find Dart/Flutter on PATH. Install Flutter by following the instruction at https://docs.flutter.dev/install.");
            }
        }
        if (updated) {
            await fs.writeFile(mcpConfigPath, JSON.stringify(mcpConfig, null, 2));
        }
    }
    catch (err) {
        utils.logWarning(`Could not configure Antigravity MCP server: ${(0, error_1.getErrMsg)(err)}`);
    }
}
async function detectAppType(rootPath) {
    try {
        await fs.access(path.join(rootPath, "pubspec.yaml"));
        return "FLUTTER";
    }
    catch {
    }
    try {
        await fs.access(path.join(rootPath, "angular.json"));
        return "ANGULAR";
    }
    catch {
    }
    try {
        const packageJsonPath = path.join(rootPath, "package.json");
        const packageJsonContent = await fs.readFile(packageJsonPath, "utf8");
        const packageJson = JSON.parse(packageJsonContent);
        const deps = { ...packageJson.dependencies, ...packageJson.devDependencies };
        if (deps.next) {
            return "NEXT_JS";
        }
        if (deps["@angular/core"]) {
            return "ANGULAR";
        }
    }
    catch {
    }
    for (const configFile of ["next.config.js", "next.config.mjs"]) {
        try {
            await fs.access(path.join(rootPath, configFile));
            return "NEXT_JS";
        }
        catch {
        }
    }
    return "OTHER";
}
const isValidFirebaseProjectId = (projectId) => {
    const projectIdRegex = /^[a-z][a-z0-9-]{4,28}[a-z0-9]$/;
    return projectIdRegex.test(projectId);
};
async function extractMetadata(rootPath, overrideProjectId) {
    const studioJsonPath = path.join(rootPath, "studio.json");
    const metadataPath = path.join(rootPath, "metadata.json");
    let metadata = {};
    for (const metadataFile of [metadataPath, studioJsonPath]) {
        try {
            const metadataContent = await fs.readFile(metadataFile, "utf8");
            metadata = JSON.parse(metadataContent);
            logger_1.logger.info(`✅ Read ${metadataFile}`);
        }
        catch (err) {
            logger_1.logger.debug(`Could not read metadata at ${metadataFile}: ${err}`);
        }
    }
    logger_1.logger.debug(`overrideProjectId ${overrideProjectId}`);
    logger_1.logger.debug(`metadata.projectId ${metadata.projectId}`);
    let projectId = overrideProjectId || metadata.projectId;
    if (!projectId) {
        try {
            const firebasercContent = await fs.readFile(path.join(rootPath, ".firebaserc"), "utf8");
            const firebaserc = JSON.parse(firebasercContent);
            projectId = firebaserc.projects?.default;
        }
        catch (err) {
            logger_1.logger.debug(`Could not read .firebaserc at ${rootPath}: ${err}`);
        }
    }
    if (projectId) {
        if (!isValidFirebaseProjectId(projectId)) {
            throw new error_1.FirebaseError(`Invalid project ID: ${projectId}.`, {
                exit: 1,
            });
        }
        logger_1.logger.info(`✅ Using Firebase Project: ${projectId}`);
    }
    else {
        logger_1.logger.debug(`❌ Failed to determine the Firebase Project ID. You can set a project later by setting the '--project' flag.`);
    }
    let appName = "firebase-studio-export";
    const blueprintPath = path.join(rootPath, "docs", "blueprint.md");
    try {
        const content = await fs.readFile(blueprintPath, "utf8");
        const nameMatch = content.match(/# \*\*App Name\*\*: (.*)/);
        if (nameMatch && nameMatch[1]) {
            appName = nameMatch[1].trim();
        }
    }
    catch (err) {
        logger_1.logger.debug(`Could not read blueprint.md at ${blueprintPath}: ${err}`);
    }
    if (appName !== "firebase-studio-export") {
        logger_1.logger.info(`✅ Detected App Name: ${appName}`);
    }
    return { projectId, appName };
}
async function updateReadme(rootPath, framework) {
    const readmePath = path.join(rootPath, "README.md");
    const readmeTemplate = await (0, templates_1.readTemplate)("firebase-studio-export/readme_template.md");
    const frameworkConfigs = {
        NEXT_JS: { startCommand: "npm run dev", localUrl: "http://localhost:9002" },
        ANGULAR: { startCommand: "npm run start", localUrl: "http://localhost:4200" },
        FLUTTER: {
            startCommand: "flutter run -d chrome --web-port=8080",
            localUrl: "http://localhost:8080",
        },
        OTHER: { startCommand: "npm run dev", localUrl: "http://localhost:9002" },
    };
    const { startCommand, localUrl } = frameworkConfigs[framework];
    let existingReadme = "";
    try {
        existingReadme = await fs.readFile(readmePath, "utf8");
    }
    catch (err) {
    }
    let newReadme = readmeTemplate
        .replace("${exportDate}", new Date().toISOString().split("T")[0])
        .replace("${startCommand}", startCommand)
        .replace("${localUrl}", localUrl);
    if (existingReadme.trim()) {
        newReadme += `\n\n---\n\n## Previous README.md contents:\n\n${existingReadme}`;
    }
    await fs.writeFile(readmePath, newReadme);
    logger_1.logger.info("✅ Updated README.md with project details and origin info");
}
async function injectAntigravityContext(rootPath, projectId, appName, nonInteractive) {
    const agentDir = path.join(rootPath, ".agents");
    const rulesDir = path.join(agentDir, "rules");
    const workflowsDir = path.join(agentDir, "workflows");
    const skillsDir = path.join(agentDir, "skills");
    await fs.mkdir(rulesDir, { recursive: true });
    await fs.mkdir(workflowsDir, { recursive: true });
    await fs.mkdir(skillsDir, { recursive: true });
    const installLocation = await prompt.select({
        message: "Where would you like to install Firebase project skills?",
        choices: [
            { name: "Locally in the project", value: "local" },
            { name: "Globally for all projects", value: "global" },
        ],
        default: "local",
        nonInteractive: nonInteractive || process.env.NODE_ENV === "test",
    });
    await (0, agentSkills_1.installAgentSkills)({
        cwd: rootPath,
        global: installLocation === "global",
        background: false,
        agentName: "gemini-cli",
        skillPackage: "firebase/agent-skills",
    });
    await (0, agentSkills_1.installAgentSkills)({
        cwd: rootPath,
        global: installLocation === "global",
        background: false,
        agentName: "gemini-cli",
        skillPackage: "genkit-ai/skills",
    });
    const systemInstructionsTemplate = await (0, templates_1.readTemplate)("firebase-studio-export/system_instructions_template.md");
    const systemInstructions = systemInstructionsTemplate.replace("${appName}", appName);
    await fs.writeFile(path.join(rulesDir, "migration-context.md"), systemInstructions);
    logger_1.logger.info("✅ Injected Antigravity rules");
    try {
        const cleanupWorkflow = await (0, templates_1.readTemplate)("firebase-studio-export/workflows/cleanup.md");
        await fs.writeFile(path.join(workflowsDir, "cleanup.md"), cleanupWorkflow);
        logger_1.logger.info("✅ Created Antigravity startup workflow");
    }
    catch (err) {
        logger_1.logger.debug(`Could not read or write startup workflow: ${(0, error_1.getErrMsg)(err)}`);
    }
}
async function getAgyCommand(startAgy) {
    if (!startAgy) {
        return undefined;
    }
    c
